File: src/main/java/net/serebryansky/common/util/LoggingUtil.java
package net.serebryansky.common.util;

import org.slf4j.MDC;
import reactor.core.publisher.Signal;

import java.util.function.Consumer;

public class LoggingUtil {
    private LoggingUtil() {
        throw new RuntimeException();
    }

    public static Consumer<Signal<?>> logOnEach(Runnable logStatement) {
        return signal -> {
            String contextValue = signal.getContextView().get("CONTEXT_KEY");
            String invoker = signal.getContextView().get("INVOKER");
            try (MDC.MDCCloseable cMdc = MDC.putCloseable("MDC_KEY", contextValue); MDC.MDCCloseable cMdc2 = MDC.putCloseable("INVOKER_KEY", contextValue)) {
                logStatement.run();
            }
        };
    }

    public static <T> Consumer<Signal<T>> logOnEach(Consumer<T> logStatement) {
        return signal -> {
            String contextValue = signal.getContextView().get("CONTEXT_KEY");
            try (MDC.MDCCloseable cMdc = MDC.putCloseable("MDC_KEY", contextValue)) {
                logStatement.accept(signal.get());
            }
        };
    }

    public static <T> Consumer<Signal<T>> logOnNext(Consumer<T> logStatement) {
        return signal -> {
            if (signal.isOnNext()) {
                String contextValue = signal.getContextView().get("CONTEXT_KEY");
                try (MDC.MDCCloseable cMdc = MDC.putCloseable("MDC_KEY", contextValue)) {
                    logStatement.accept(signal.get());
                }
            }
        };
    }
}

File: src/main/kotlin/net/serebryansky/common/util/LoggingUtil.kt
package net.serebryansky.common.util

import org.slf4j.MDC
import reactor.core.publisher.Signal
import java.util.function.Consumer

class LoggingUtil private constructor() {
    init {
        throw RuntimeException()
    }

    companion object {
        fun logOnEach(logStatement: Runnable): Consumer<Signal<*>> {
            return Consumer { signal: Signal<*> ->
                val contextValue = signal.contextView.get<String>("CONTEXT_KEY")
                val invoker = signal.contextView.get<String>("INVOKER")
                MDC.putCloseable("INVOKER_KEY", invoker).use { cMdc2 -> MDC.putCloseable("MDC_KEY", contextValue).use { cMdc -> logStatement.run() } }
            }
        }

        @JvmStatic
        fun <T> logOnEach(logStatement: Consumer<T?>): Consumer<Signal<T?>> {
            return Consumer { signal: Signal<T?> ->
                val contextValue = signal.contextView.get<String>("CONTEXT_KEY")
                val invoker = signal.contextView.get<String>("INVOKER")
                MDC.putCloseable("INVOKER_KEY", invoker).use { cMdc2 -> MDC.putCloseable("MDC_KEY", contextValue).use { cMdc -> logStatement.accept(signal.get()) } }
            }
        }

        fun <T> logOnNext(logStatement: Consumer<T>): Consumer<Signal<T>> {
            return Consumer { signal: Signal<T> ->
                if (signal.isOnNext) {
                    val contextValue = signal.contextView.get<String>("CONTEXT_KEY")
                    val invoker = signal.contextView.get<String>("INVOKER")
                    MDC.putCloseable("INVOKER_KEY", invoker).use { cMdc2 -> MDC.putCloseable("MDC_KEY", contextValue).use { cMdc -> logStatement.accept(signal.get()!!) } }
                }
            }
        }
    }
}
